<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 17 Metageomics | 生物信息随记</title>
<meta name="author" content="Hua Zou">
<meta name="description" content="本文聚焦用生物信息学的方法挖掘宏基因组数据背后的生物学意义。 从环境中提取包含所有微生物的DNA，然后建库测序获取fastq等二代测序数据，生信是指对这些数据进行处理和提取它们背后的生物学意义，解析出对环境影响产生相应的微生物。 宏基因组数据分析的优劣点 相关数据分析工具  17.1 宏基因组介绍 metagenomics通常有多种含义，它包含 基于marker...">
<meta name="generator" content="bookdown 0.34 with bs4_book()">
<meta property="og:title" content="Chapter 17 Metageomics | 生物信息随记">
<meta property="og:type" content="book">
<meta property="og:url" content="https://zouhua.top/DraftNotes/Metageomics.html">
<meta property="og:description" content="本文聚焦用生物信息学的方法挖掘宏基因组数据背后的生物学意义。 从环境中提取包含所有微生物的DNA，然后建库测序获取fastq等二代测序数据，生信是指对这些数据进行处理和提取它们背后的生物学意义，解析出对环境影响产生相应的微生物。 宏基因组数据分析的优劣点 相关数据分析工具  17.1 宏基因组介绍 metagenomics通常有多种含义，它包含 基于marker...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 17 Metageomics | 生物信息随记">
<meta name="twitter:description" content="本文聚焦用生物信息学的方法挖掘宏基因组数据背后的生物学意义。 从环境中提取包含所有微生物的DNA，然后建库测序获取fastq等二代测序数据，生信是指对这些数据进行处理和提取它们背后的生物学意义，解析出对环境影响产生相应的微生物。 宏基因组数据分析的优劣点 相关数据分析工具  17.1 宏基因组介绍 metagenomics通常有多种含义，它包含 基于marker...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><link href="libs/Poppins-0.4.6/font.css" rel="stylesheet">
<script src="libs/bs3compat-0.6.0/transition.js"></script><script src="libs/bs3compat-0.6.0/tabs.js"></script><script src="libs/bs3compat-0.6.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="libs/htmlwidgets-1.6.2/htmlwidgets.js"></script><link href="libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="libs/datatables-binding-0.31/datatables.js"></script><link href="libs/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="libs/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="libs/dt-core-1.13.6/js/jquery.dataTables.min.js"></script><link href="libs/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet">
<script src="libs/crosstalk-1.2.0/js/crosstalk.min.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<style type="text/css">
    /* Used with Pandoc 2.11+ new --citeproc when CSL is used */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
        }
    .hanging div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }
  </style>
<link rel="stylesheet" href="assets/style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">生物信息随记</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> 介绍</a></li>
<li class="book-part">Data Set</li>
<li><a class="" href="ZeybelDataset.html"><span class="header-section-number">2</span> Zeybel Dataset</a></li>
<li class="book-part">Statistical Methods</li>
<li><a class="" href="GEEandMLM.html"><span class="header-section-number">3</span> GEE and MLM</a></li>
<li><a class="" href="HypothesisTestingMethods.html"><span class="header-section-number">4</span> Hypothesis Testing Methods</a></li>
<li><a class="" href="BatchEffectCorrection.html"><span class="header-section-number">5</span> Batch Effect Correction</a></li>
<li><a class="" href="LinearModelonMicrobialCommunity.html"><span class="header-section-number">6</span> Linear Model on Microbial Community</a></li>
<li><a class="" href="PermutationTest.html"><span class="header-section-number">7</span> Permutation Test</a></li>
<li><a class="" href="SurvivalAnalysis.html"><span class="header-section-number">8</span> Survival Analysis</a></li>
<li class="book-part">Metabolomics Data Analysis</li>
<li><a class="" href="dataprocessing.html"><span class="header-section-number">9</span> Data Processing</a></li>
<li><a class="" href="DimensionReduction.html"><span class="header-section-number">10</span> Dimension Reduction</a></li>
<li><a class="" href="DifferetialAnalysis.html"><span class="header-section-number">11</span> Differetial Analysis</a></li>
<li><a class="" href="FunctionalAnalysis.html"><span class="header-section-number">12</span> Functional Analysis</a></li>
<li><a class="" href="MetOriginAnalysis.html"><span class="header-section-number">13</span> MetOrigin Analysis</a></li>
<li><a class="" href="OtherAnalysis.html"><span class="header-section-number">14</span> Other Analysis</a></li>
<li class="book-part">Machine Learning</li>
<li><a class="" href="randomforestalgorithm.html"><span class="header-section-number">15</span> Random Forest Algorithm</a></li>
<li><a class="" href="XGBoostalgorithm.html"><span class="header-section-number">16</span> XGBoost Algorithm</a></li>
<li class="book-part">Knowledge</li>
<li><a class="active" href="Metageomics.html"><span class="header-section-number">17</span> Metageomics</a></li>
<li><a class="" href="Notes.html"><span class="header-section-number">18</span> Notes</a></li>
<li><a class="" href="references.html">References</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/HuaZou/DraftNotes">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="Metageomics" class="section level1" number="17">
<h1>
<span class="header-section-number">17</span> Metageomics<a class="anchor" aria-label="anchor" href="#Metageomics"><i class="fas fa-link"></i></a>
</h1>
<p>本文聚焦用生物信息学的方法挖掘宏基因组数据背后的生物学意义。</p>
<p>从环境中提取包含所有微生物的DNA，然后建库测序获取fastq等二代测序数据，生信是指对这些数据进行处理和提取它们背后的生物学意义，解析出对环境影响产生相应的微生物。</p>
<ul>
<li><p>宏基因组数据分析的优劣点</p></li>
<li><p>相关数据分析工具</p></li>
</ul>
<div id="宏基因组介绍" class="section level2" number="17.1">
<h2>
<span class="header-section-number">17.1</span> 宏基因组介绍<a class="anchor" aria-label="anchor" href="#%E5%AE%8F%E5%9F%BA%E5%9B%A0%E7%BB%84%E4%BB%8B%E7%BB%8D"><i class="fas fa-link"></i></a>
</h2>
<p>metagenomics通常有多种含义，它包含</p>
<ul>
<li><p>基于marker genes测序，如我们常说的16s rRNA基因测序或者真菌的ITS测序（扩增子测序范畴）</p></li>
<li><p>shotgun metagenome测序，也就是所有DNA都采用鸟枪法打断测序（通过DNA insert size筛选测序片段）</p></li>
<li><p>另一种是meta-transcriptome测序（宏转录组）</p></li>
</ul>
<p>不同测序方法的目的是不同的，</p>
<ul>
<li><p>比如仅仅想知道环境中有什么物种存在，采用标记基因16s rRNA DNA测序即可 （<strong>16s metagenomics</strong>）</p></li>
<li><p>比如想知道环境中有什么物种存在且它们发挥什么功能，采用shotgun metagenome测所有DNA即可 (<strong>Metagenomics</strong>)</p></li>
<li><p>但是如果想知道当前环境中微生物正在发挥什么功能，则采用检测mRNA，蛋白质或代谢组即可 (<strong>Metatranscriptomics, Metaproteomics and Metametabolomics</strong>)</p></li>
</ul>
<div class="inline-table"><table style="width:99%;" class="table table-sm">
<colgroup>
<col width="20%">
<col width="50%">
<col width="28%">
</colgroup>
<thead><tr class="header">
<th>技术</th>
<th>优点和缺点</th>
<th>应用</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>扩增子测序（16s, ITS）</td>
<td>
<p>+快速低成本识别大多数细菌和真核生物</p>
<p>-不能捕获其它非靶点基因</p>
<p>-扩增子偏好性（PCR过程中存在，三代测序全长16s rRNA基因可以避免该问题）</p>
<p>-不能捕获病毒</p>
</td>
<td>
<p>表征环境中微生物的存在</p>
<p>微生态学</p>
<p>基于rRNA标记基因的进化树研究</p>
</td>
</tr>
<tr class="even">
<td>MGS（shotgun）of DNA or RNA</td>
<td>
<p>+没有偏好性</p>
<p>+可检测细菌、古菌、病毒和真核生物</p>
<p>+可从头组装微生物组</p>
<p>-需大量reads</p>
<p>-reads可能来自于宿主</p>
<p>-物种分类需参考数据库</p>
</td>
<td>
<p>可知晓环境的当前多个类型微生物存在情况</p>
<p>功能基因组学研究</p>
<p>病原微生物检测</p>
<p>系统进化树</p>
</td>
</tr>
<tr class="odd">
<td>Meta-transcriptomics (mRNA)</td>
<td>
<p>+识别活跃状态下的基因和通路</p>
<p>-mRNA是不稳定的易降解</p>
<p>-多个纯化和扩增mRNA的步骤容易引入较多的噪音</p>
</td>
<td>当前环境活跃状态的转录组</td>
</tr>
</tbody>
</table></div>
<p>挑战：</p>
<ul>
<li><p>生态系统的复杂性（微生物数目和种类繁多，相互关系复杂）</p></li>
<li><p>数据库的完整性（较多微生物缺乏参考数据库）</p></li>
<li><p>测序深度（稀少微量的微生物无法检测出来）</p></li>
<li><p>计算资源（微生物从头组装等需要消耗巨量的计算资源）</p></li>
</ul>
<div id="宏基因组常用分析流程" class="section level3" number="17.1.1">
<h3>
<span class="header-section-number">17.1.1</span> 宏基因组常用分析流程<a class="anchor" aria-label="anchor" href="#%E5%AE%8F%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%B8%B8%E7%94%A8%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B"><i class="fas fa-link"></i></a>
</h3>
<p>参考 A review of methods and databases for metagenomic classification and assembly (Ref:)<span class="citation"><a class="footnote-ref" tabindex="0" data-toggle="popover" data-content="&lt;p&gt;Florian P Breitwieser, Jennifer Lu, and Steven L Salzberg, &lt;span&gt;“A Review of Methods and Databases for Metagenomic Classification and Assembly,”&lt;/span&gt; &lt;em&gt;Briefings in Bioinformatics&lt;/em&gt; 20, no. 4 (2019): 1125–36.&lt;/p&gt;"><sup>3</sup></a></span></p>
<div class="figure" style="text-align: center">
<span style="display:block;" id="fig:unnamed-chunk-2"></span>
<img src="InputData/figures/common_MGS.png" alt="Common analysis procedures for metagenomics data" width="50%" height="50%"><p class="caption">
Figure 4.1: Common analysis procedures for metagenomics data
</p>
</div>
<p><strong>Figure1</strong>. Common analysis procedures for metagenomics data. Note that the order of some of the analysis steps can be shuffled. For example, reads might be binned before assembly or before taxonomic assignment, so that the downstream algorithms can work only with a subset of the data.</p>
<p>上述步骤分为：</p>
<ul>
<li><p>fastq质量控制：去除低质量的reads（FastQC, fastp, Fastq_screen, BBtools, Trimmomatic, Cutadapt, khmer/diginorm, MultiQC）</p></li>
<li><p>metagenome组装：从头组装reads成基因组 (微生物比对分类等软件：Kraken, CLARK, Kallisto, Kaiju, GOTTCHA, Centrifuge, Metaphlan2/3, mOTU, Mash, sourmash; 微生物组装：Megahit, SPAdes, MetaSPAdes, IDBA-UD, MOCAT2)</p></li>
<li><p>binning分箱：组装后的分箱，获得精度更高的微生物组 (COCACOLA, CONCOCT, MetaBAT, VizBin)</p></li>
<li><p>物种注释：将得到的基因组草图注释到参考数据库获得物种分类信息 (质量控制：BUSCO, CheckM)</p></li>
<li><p>下游分析</p></li>
</ul>
<p>这些步骤涉及软件众多。</p>
</div>
</div>
<div id="reference-14" class="section level2" number="17.2">
<h2>
<span class="header-section-number">17.2</span> Reference<a class="anchor" aria-label="anchor" href="#reference-14"><i class="fas fa-link"></i></a>
</h2>
<ul>
<li><p>Breitwieser, Florian P., Jennifer Lu, and Steven L. Salzberg. “A review of methods and databases for metagenomic classification and assembly.” Briefings in bioinformatics 20.4 (2019): 1125-1136.</p></li>
<li><p><a href="https://trainings.migale.inrae.fr/posts/2022-06-20-module24/content/slides.html#15" class="uri">https://trainings.migale.inrae.fr/posts/2022-06-20-module24/content/slides.html#15</a></p></li>
</ul>
</div>
</div>

  <div class="chapter-nav">
<div class="prev"><a href="XGBoostalgorithm.html"><span class="header-section-number">16</span> XGBoost Algorithm</a></div>
<div class="next"><a href="Notes.html"><span class="header-section-number">18</span> Notes</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#Metageomics"><span class="header-section-number">17</span> Metageomics</a></li>
<li>
<a class="nav-link" href="#%E5%AE%8F%E5%9F%BA%E5%9B%A0%E7%BB%84%E4%BB%8B%E7%BB%8D"><span class="header-section-number">17.1</span> 宏基因组介绍</a><ul class="nav navbar-nav"><li><a class="nav-link" href="#%E5%AE%8F%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%B8%B8%E7%94%A8%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B"><span class="header-section-number">17.1.1</span> 宏基因组常用分析流程</a></li></ul>
</li>
<li><a class="nav-link" href="#reference-14"><span class="header-section-number">17.2</span> Reference</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/HuaZou/DraftNotes/blob/main/80-KW-metagenomics.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/HuaZou/DraftNotes/edit/main/80-KW-metagenomics.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>生物信息随记</strong>" was written by Hua Zou. It was last built on 2023-08-24 and updated on 2024-02-06.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
